### PR TITLE
add lower and higher interpolation to quantiles

### PR DESCRIPTION
**Pull request by marcvanheerden**
Add interpolation options to quantiles, initially this PR is a draft to get early feedback.
Just added Lower and Higher interpolation and only in Rust.

TODO:
   implement changes in python
   add other interpolation options to quantile and median.

### CODE REVIEW

**Comment by ritchie46** [comment]
Looks good @marcvanheerden. Its quite a lot of dispatch code, sorry about that. :/

**Comment by marcvanheerden** [comment]
sorry for the delay, struggling to get the builds to pass. Getting different outputs from black locally vs in the builds. will keep trying to figure it out

**Comment by ritchie46** [comment]
> sorry for the delay, struggling to get the builds to pass. Getting different outputs from black locally vs in the builds. will keep trying to figure it out

Make sure that the black version is correct. The output differs between versions.

Did you run `make pre-commit`?

#2025 will likely also make this easier.

**Comment by ritchie46** [comment]
@marcvanheerden if you want I can also do the linting for you.

**Comment by marcvanheerden** [comment]
Thanks! I want to write some more tests and take the PR out of draft tomorrow. Let me know if you're in a rush to get this done

File: py-polars/polars/internals/expr.py [hunk 1]
```diff
@@ -1077,11 +1077,11 @@ def is_duplicated(self) -> "Expr":
         """
         return wrap_expr(self._pyexpr.is_duplicated())
 
-    def quantile(self, quantile: float) -> "Expr":
+    def quantile(self, quantile: float, interpolation: str = "nearest") -> "Expr":
         """
         Get quantile value.
         """
```
Comment (ritchie46) [inline py-polars/polars/internals/expr.py:1083]: Maybe add parameters and the possible interpolation strings to the docstring.

File: py-polars/polars/internals/frame.py [hunk 2]
```diff
@@ -3303,7 +3303,7 @@ def median(self) -> "DataFrame":
         """
         return wrap_df(self._df.median())
 
-    def quantile(self, quantile: float) -> "DataFrame":
+    def quantile(self, quantile: float, interpolation: str = "nearest") -> "DataFrame":
```
Comment (ritchie46) [inline py-polars/polars/internals/frame.py:3306]: Let's add a parameters sections which explains the possible interpolation strings

File: py-polars/polars/internals/frame.py [hunk 3]
```diff
@@ -4251,13 +4251,15 @@ def n_unique(self) -> DataFrame:
             return wrap_df(self._df.downsample(self.by, self.rule, self.n, "n_unique"))
         return wrap_df(self._df.groupby(self.by, self.selection, "n_unique"))
 
-    def quantile(self, quantile: float) -> DataFrame:
+    def quantile(self, quantile: float, interpolation: str = "nearest") -> DataFrame:
```
Comment (ritchie46) [inline py-polars/polars/internals/frame.py:4254]: Let's add a parameters sections which explains the possible interpolation strings

File: py-polars/polars/internals/frame.py [hunk 4]
```diff
@@ -4072,11 +4072,11 @@ def n_unique(self) -> DataFrame:
         """
         return self._select_all().n_unique()
 
-    def quantile(self, quantile: float) -> DataFrame:
+    def quantile(self, quantile: float, interpolation: str = "nearest") -> DataFrame:
```
Comment (ritchie46) [inline py-polars/polars/internals/frame.py:4075]: Let's add a parameters sections which explains the possible interpolation strings

File: py-polars/polars/internals/series.py [hunk 5]
```diff
@@ -844,7 +844,7 @@ def median(self) -> float:
         """
         return self._s.median()
 
-    def quantile(self, quantile: float) -> float:
+    def quantile(self, quantile: float, interpolation: str = "nearest") -> float:
```
Comment (ritchie46) [inline py-polars/polars/internals/series.py:847]: Let's add a parameters sections which explains the possible interpolation strings

**Comment by ritchie46** [comment]
Looks good @marcvanheerden thanks a lot! 

I only commented on some places where we miss a docstring that explains which `str` values we can pass to the `interpolation` keyword.

**Comment by marcvanheerden** [comment]
thanks @ritchie46 !

**Comment by marcvanheerden** [comment]
Should be ready now, let me know if you want any changes. The default interpolation doesn't behave exactly the way "nearest" does in Pandas but I've kept it like it was in polars before because I think it is more correct like this. 

having the output in the original types means that for integer types the "midpoint" and "linear" interpolation options essentially do the same thing as "lower". Let me know if you want to change this?

Also, it might be useful to have another quantile function that always returns f64 like .median() does. 

Happy to make any changes you want

**Comment by ritchie46** [comment]
> be useful to have another quantile function that always returns f64 like .median() does.
> 
> Happy to make any changes you want

Again, thanks a lot!

> having the output in the original types means that for integer types the "midpoint" and "linear" interpolation options essentially do the same thing as "lower". Let me know if you want to change this?

Not a blocker for this PR. But I think it makes sense to always return `f64` indeed (except if the dtype is `f32`).

**Pull request merged**